crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+     * doesn't require more - If it does, we'll fail.
+     */
+    ciphertext = OPENSSL_malloc(plainlen + taglen);
+    if (ciphertext == NULL) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Create and initialise the context */
+    if (!(ctx = EVP_CIPHER_CTX_new())) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Initialise the encryption operation. */
+    enc = EVP_CIPHER_fetch(libctx, hpke_aead_tab[aead_ind].name, propq);
+    if (enc == NULL) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (EVP_EncryptInit_ex(ctx, enc, NULL, NULL, NULL) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    EVP_CIPHER_free(enc);
+    enc = NULL;
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, ivlen, NULL) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Initialise key and IV */
+    if (EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Provide any AAD data. */
+    if (aadlen != 0 && aad != NULL) {
+        if (EVP_EncryptUpdate(ctx, NULL, &len, aad, aadlen) != 1) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    /*
+     * Provide the message to be encrypted, and obtain the encrypted output.
+     * EVP_EncryptUpdate can be called multiple times if necessary
+     */
+    if (EVP_EncryptUpdate(ctx, ciphertext, &len, plain, plainlen) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ciphertextlen = len;
+    /* Finalise the encryption. */
+    if (EVP_EncryptFinal_ex(ctx, ciphertext + len, &len) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ciphertextlen += len;
+    /*
+     * Get the tag This isn't a duplicate so needs to be added to the ciphertext
+     */
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, taglen, tag) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    memcpy(ciphertext + ciphertextlen, tag, taglen);
+    ciphertextlen += taglen;
+    if (ciphertextlen > *cipherlen) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    *cipherlen = ciphertextlen;
+    memcpy(cipher, ciphertext, ciphertextlen);
+
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(enc);
+    OPENSSL_free(ciphertext);
+    return erv;
+}
+/*
+ * @brief map a kem_id and a private key buffer into an EVP_PKEY
+ *
+ * Note that the buffer is expected to be some form of the encoded
```

Comment:
Are all these combinations really necessary, this seems overly complicated.

Reply:
Fair question. My take is that it's better to support decoding inputs liberally if one can do that safely, on the basis that the HPKE code can save callers writing more code munging private key encodings. But one could well argue that only supporting a single input format might be "cleaner" in terms of fewer lines of code within the library. There's also a reasonable argument wrt maintaining consistency with other APIs. I'm open to change but the current setup has I think made it easier to test ECH with various clients and servers where differently formatted inputs are likely.

Reply:
I think that then becomes a test utility, rather than becoming embedded in the implementation..

Reply:
so this is overtaken by events - that code's no longer needed here, but something like it may be needed for ECH, will find that out in next day or so